src/main/java/com/powsybl/openloadflow/ac/nr/NewtonRaphson.java

```diff
@@ -63,7 +65,36 @@ public static List<Pair<Equation<AcVariableType, AcEquationType>, Double>> findL
                 .collect(Collectors.toList());
     }
 
-    private NewtonRaphsonStatus runIteration(StateVectorScaling svScaling, MutableInt iterations) {
+    public static void reportLargestMismatch(Reporter reporter, EquationSystem<AcVariableType, AcEquationType> equationSystem, double[] mismatch, LfNetwork network, int iteration) {
+        List<Pair<Equation<AcVariableType, AcEquationType>, Double>> mismatchEquations = equationSystem.getIndex().getSortedEquationsToSolve().stream()
+                .map(equation -> Pair.of(equation, mismatch[equation.getColumn()]))
+                .filter(e -> Math.abs(e.getValue()) > Math.pow(10, -7))
+                .sorted(Comparator.comparingDouble((Map.Entry<Equation<AcVariableType, AcEquationType>, Double> e) -> Math.abs(e.getValue())).reversed())
+                .collect(Collectors.toList());
+
+        List<AcEquationType> acEquationTypes = List.of(AcEquationType.BUS_TARGET_P, AcEquationType.BUS_TARGET_Q, AcEquationType.BUS_TARGET_V);
+
+        for (AcEquationType acEquationType : acEquationTypes) {
+            mismatchEquations.stream()
+                    .filter(e -> e.getKey().getType() == acEquationType)
+                    .limit(1)
+                    .forEach(e -> {
```

Comment:
If you use the map you won't need this anymore, but don't do `limit(1).forEach()`, it would be clearer to use `findFirst().ifPresent()` instead.

Reply:
Thank you @flo-dup  for the advice!